### DISCUSSION TITLE
How does Next 14 indicate a response is finished?

### DISCUSSION THREAD

**Discussion by blairmcalpine** [vercel/next.js#59172] [Help]
### Summary

We have a custom Next server that uses `getRequestHandler` inside of a node server to render pages. This server uses gRPC for communication, meaning we need to adapt the gRPC stream to an HTTP request and response for NextJS to be able to handle the request. Before NextJS 14, we relied on the `res.end()` function to be called in order to know when to indicate the stream was finished. As of upgrading to NextJS 14, the `res.end()` function is no longer called.

When testing this same implementation with an express server and passing it the provided `req` and `res` objects, things work as expected and `res.end()` is called. My hypothesis is that NextJS 14 relies on some property of `req` or `res` that our "fake" gRPC to HTTP adaptations don't have, and this causes `res.end()` to never be called.

TL;DR Is what properties of `req` and `res` does the Next request handler rely on in NextJS 14 that it didn't rely on before?

It is worth noting as well that our "fake" gRPC methods implement all required attributes of `request` and `response` according to the `http` package types.
